Implement ID*

djinnome opened the pull request:
Closes #69 by implementing the ID* algorithm.

This PR also adds notebooks that explain the concepts behind the ID and IDC algorithm and walks through its execution

See the notebooks:
- Identification https://nbviewer.jupyter.org/github/y0-causal-inference/y0/blob/identification/notebooks/Identification.ipynb
- Counterfactual Identification https://github.com/y0-causal-inference/y0/blob/identification/notebooks/Counterfactual%20Identification.ipynb

cthoyt commented:
@djinnome please resolve conflicts with master so I can give this a review

cthoyt commented:
Other feedback on the notebook itself:
- could you use labels to go with each graph that's output? I've never seen the `display()` function before but since you're using some matplotlib stuff, it should be possible
- reformat code with black (https://pypi.org/project/black-nb/)
- type annotations on the functions for theorem 2
- the ID* and ICD* section is a bit of a stub. maybe remove from this PR and elaborate in a future one?
- Resize the images, it's a big jarring to scroll through this with them at maximum width size

djinnome commented:
Is it time to figure out why all the lint and documentation tests don't pass?

- [ ] 

> @djinnome please resolve conflicts with master so I can give this a review

Done.



> Other feedback on the notebook itself:
> 
> * could you use labels to go with each graph that's output? I've never seen the `display()` function before but since you're using some matplotlib stuff, it should be possible

`display()` is an [IPython thing](https://ipython.readthedocs.io/en/stable/api/generated/IPython.display.html#IPython.display.display).
Adding a title to each graph was a cinch thanks to your `draw()` function.  I also added an `expression` property to the `Query` data class to facilitate visualization.   

> * reformat code with black (https://pypi.org/project/black-nb/)
I don't think you can reformat jupyter notebook code. 
```bash
$ black notebooks
No Python files are present to be formatted. Nothing to do 😴
```

> * type annotations on the functions for theorem 2

Done! And I cleaned up the Markdown to use boolean variables in the discussion.

> * the ID* and ICD* section is a bit of a stub. maybe remove from this PR and elaborate in a future one?
Done!
> * Resize the images, it's a big jarring to scroll through this with them at maximum width size
Done!

cthoyt commented:
I gave the wrong black link, it's actually https://pypi.org/project/nb-black/. I applied it.
The credit for the draw() function completely goes to Joseph! I'm not going to tag him so he can keep his sanity :)

We need to clean up the code graveyard before merge - both the y0.algorithm.identify.old and all of the commented examples in the bottom of the tests/test_algorithm/test_id_alg.py need to be deleted.

It also looks like you added a lot more examples. Almost all of them (except figure 6A) don't actually seem relevant to this PR, so do you think you could split them into a new one? I'd like to discuss the formalism since these are the first examples with counterfactuals inside them and we have to decide how we want to be able to convert back and forth. Definitely not a discussion that I want to hold up this PR for

djinnome commented:
Hmmm... the code graveyard at the bottom of `test_id_alg.py` actuallly should be resurrected now that we can produce an estimand from the identify algorithm.  The challenge will be confirming it outputs the right estimand, but by comparing it with the output of causaleffect (without simplification, with canonicalization), we should be able to confirm correctness.

cthoyt commented on src/y0/algorithm/identify/id_star.py:
```diff
@@ -0,0 +1,233 @@
+    )
+    for node in graph.topological_sort():
+        for intervention in worlds:
+            if (
```
make this conditional an auxiliary function and give some docs as to what's getting checked

cthoyt commented on src/y0/algorithm/identify/id_star.py:
```diff
@@ -0,0 +1,233 @@
+            for treatments_from_world_1, treatments_from_world_2 in combinations(worlds, 2)
+            for u in graph.nodes()
+            # Don't add an edge if a variable is intervened on in either world.
+            if (u not in treatments_from_world_1)
```
is this logic correct? the comment says in either world, but it's 4 conditions and'd together

cthoyt replied:
this check is done more than once, maybe better to make an auxiliary function for it too

djinnome replied:
The logic is correct. If u and v have a bidirected edge in the observed world, they have a bidirected edge in a counterfactual world **unless** either have been intervened upon.  There are two ways an intervention could happen. Either intervening to have the same value as in the observed world, or a different value.  I do like the idea of refactoring the duplicate code

cthoyt commented on src/y0/algorithm/taheri_design.py:
```diff
@@ -200,7 +200,7 @@ def _get_result(
 
 
-def iterate_lvdags(
+def iterate_lvoodags(
```
typo?

cthoyt commented on src/y0/graph.py:
```diff
@@ -414,6 +415,15 @@ def is_connected(self) -> bool:
 
+    def relabel_nodes(self, mapping: Mapping[CounterfactualVariable, Variable]) -> NxMixedGraph:
+        directed = relabel_nodes(self.directed, mapping)
+        undirected = relabel_nodes(self.directed, mapping)
```
should be `self.undirected`?

robertness commented:
Holy cow guys.  Merge this branch!

cthoyt commented on src/y0/algorithm/identify/id_star.py:
```diff
@@ -0,0 +1,383 @@
+
+def get_district_event(district: District, interventions: Set[Intervention]) -> Event:
+    return {
+        # FIXME is node.get_base() right? This does not create interventions but rather variables
```
@djinnome please advise

cthoyt commented on src/y0/algorithm/identify/cg.py:
```diff
@@ -0,0 +1,318 @@
+    return combine_worlds(graph, world_graphs, worlds)
+
+
+def _stitch_1(graph: NxMixedGraph, worlds: Worlds):
```
@djinnome I refactored these 4 stitching operations out of the main `combine_worlds()` function. Can you please give some high level narration for each of them as docstrings?
